### ISSUE TITLE
Enable autodiff to be used with custom number systems types

### ISSUE THREAD

**Issue by Ravenwater** [autodiff/autodiff#91]
We would like to use the autodiff library with custom arithmetic types, such as posits, valids, unums, Elias, fixed-point, etc. so that autodiff can be used in environments where these types are instrumental to the computational fidelity or efficiency.

**Comment by allanleal** [comment]
Great, thanks @Ravenwater for bringing this issue to here. I have a big PR #69 where I remember having fixed some issues, but not so sure. The PR is 99.9% ready and it introduces new capabilities and a new type `Real<N, T>`, where `N` is the order, and `T` is the floating point type - this is a tailor-made autodiff type for high-order directional derivatives and Taylor series of single and multivariable functions along a direction. Anyway, it's in our interest that any numeric type be supported together with `autodiff` types, such as the ones you mention above.

**Comment by Ravenwater** [comment]
@allanleal looking forward to it. Hoping to learn from autodiff what type of number traits api you'll need to use these types.

**Comment by allanleal** [comment]
In the PR #69 , you can have a look at this file [`numbertraits.hpp`](https://github.com/allanleal/autodiff/blob/improve-high-order-derivatives-using-array/autodiff/common/numbertraits.hpp) and in [`meta.hpp`](https://github.com/allanleal/autodiff/blob/improve-high-order-derivatives-using-array/autodiff/common/meta.hpp). But to summarize, what we'll have is more or less:

~~~c++
template<typename T>
constexpr bool isNumber = std::is_arithmetic<PlainType<T>>::value;

/// An auxiliary template type to indicate NumberTraits has not been defined for a type.
template<typename T>
struct NumericTypeInfoNotDefinedFor { using type = T; };

/// A number traits to be defined for each autodiff number.
template<typename T>
struct NumberTraits
{
    /// The underlying floating point type of the autodiff number type.
    using NumericType = std::conditional_t<isNumber<T>, T, NumericTypeInfoNotDefinedFor<T>>;

    /// The order of the autodiff number type.
    static constexpr auto Order = 0;
};

/// A template alias to get the underlying floating point type of an autodiff number.
template<typename T>
using NumericType = typename NumberTraits<PlainType<T>>::NumericType;
~~~

It looks like we need to make `isNumber` less dependent on `std::is_arithmetic`.

**Comment by Ravenwater** [comment]
It looks like we need to make `isNumber` less dependent on `std::is_arithmetic`.

this will require another layer of indirection, :-), to facilitate specialization of an is_arithmetic concept for custom types

**Comment by allanleal** [comment]
Agree. I'll try to use your Universal Number Arithmetic (UNA) library in that PR in the next days and see how to get this working. It would then be exciting to have some examples that combine autodiff and UNA in the tutorials section of [autodiff.github.io](autodiff.github.io). I think @supersega is also investigating this somehow.

**Comment by Ravenwater** [comment]
@allanleal @supersega awesome!  One of the big reasons for moving away from IEEE is numerical consistency of computational results. The other reasons, which are driven by the DL folks, is performance and energy efficiency.

A good set of tests would involve gradient searches and the impact precision and dynamic range have on the quality of the solution. This might be a bigger task then is reasonable for tutorial level documentation/examples, but that is why I have come to autodiff, :-)

Let me know what you need, and I will oblige: this is very exciting.

**Comment by supersega** [comment]
@Ravenwater Please wait until #92 merged. I believe it resolve your problem.

**Comment by supersega** [comment]
Hi @Ravenwater, please check that autodiff works with posit.
